featurebyte/common/utils.py

```diff
@@ -119,22 +120,54 @@ def dataframe_to_arrow_bytes(dataframe: pd.DataFrame) -> bytes:
     return data
 
 
-def dataframe_from_arrow_stream(buffer: Any) -> pd.DataFrame:
+def dataframe_from_arrow_stream(buffer: Any, num_rows: int = 0) -> pd.DataFrame:
     """
     Read data from arrow byte stream to pandas dataframe
 
     Parameters
     ----------
     buffer: Any
         buffer-like
+    num_rows: int
+        Number of rows to read
 
     Returns
     -------
     pd.DataFrame
         Pandas Dataframe object
     """
     reader = pa.ipc.open_stream(buffer)
-    return reader.read_all().to_pandas()
+    if num_rows == 0:
+        arrow_table = reader.read_all()
+    else:
+        batches = []
+        batches.append(reader.read_next_batch())
+        try:
+            with alive_bar(
+                total=num_rows,
+                title="Downloading table",
+                **get_alive_bar_additional_params(),
+            ) as progress_bar:
```

Comment:
Is this `dataframe_from_arrow_stream` function only used by API objects but not in the services? If so would be good to move it to under the `api` directory to make that clear. I think using `alive_bar` in backend service code might cause issues due to the use of threading etc.

Reply:
Ya, i also encountered `dictionary changed size during iteration` before when using `alive_bar`. Would be good to move it to `api` specific directory to prevent people use it in backend service.

Reply:
The function is used in both places currently but alive bar is only used from the api (by passing in num_rows) to display download progress. I can refactor this into 2 different functions and move one under api